File: tools/reference-implementations/python/vlp.py
"""Python reference implementation of VLP v0.2 base frame.

Non-normative. The authoritative specification is at
`book/src/spec/vlp.md`. This module exists so an external reader can
confirm their understanding of the byte layout against working code
without needing a Rust toolchain.

Requires Python 3.8+. Standard library only.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass
from enum import IntEnum


MAGIC = b"\x56\x41"  # "VA"
VERSION = 0x02
NONCE_TERMINAL = 0xFFFFFFFFFFFFFFFF


class Status(IntEnum):
    OK = 0
    DEGRADED = 1
    CRITICAL = 2
    STALL = 3  # observer-synthesized only — MUST NOT appear on the wire


STATUS_BY_NAME = {
    "ok": Status.OK,
    "degraded": Status.DEGRADED,
    "critical": Status.CRITICAL,
}


class DecodeError(Exception):
    """Raised on any wire-format validation failure.

    The `kind` attribute is the spec-defined error variant name, which
    matches the strings in `tools/vlp-test-vectors.json`:

        BadMagic, BadVersion, BadCrc, BadStatus, StallOnWire,
        BadPid, BadTimestamp, BadNonce.
    """

    def __init__(self, kind: str, detail: str = ""):
        super().__init__(f"{kind}: {detail}" if detail else kind)
        self.kind = kind


# ---------------------------------------------------------------------------
# CRC-32C (Castagnoli)
# ---------------------------------------------------------------------------

_POLY_REFLECTED = 0x82F63B78


def _build_crc_table() -> tuple[int, ...]:
    table = []
    for i in range(256):
        c = i
        for _ in range(8):
            c = (c >> 1) ^ _POLY_REFLECTED if c & 1 else c >> 1
        table.append(c)
    return tuple(table)


_CRC_TABLE = _build_crc_table()


def crc32c(data: bytes) -> int:
    """Compute the CRC-32C (Castagnoli) checksum of *data*.

    Reflected polynomial 0x82F63B78, init 0xFFFFFFFF, refin/refout,
    xorout 0xFFFFFFFF. Matches RFC 3720 appendix B.
    """
    crc = 0xFFFFFFFF
    for b in data:
        crc = _CRC_TABLE[(crc ^ b) & 0xFF] ^ (crc >> 8)
    return crc ^ 0xFFFFFFFF


# ---------------------------------------------------------------------------
# Frame
# ---------------------------------------------------------------------------


@dataclass(frozen=True)
class Frame:
    status: Status
    pid: int
    timestamp: int
    nonce: int
    payload: int


def encode(
    status: int | str | Status,
    pid: int,
    timestamp: int,
    nonce: int,
    payload: int,
) -> bytes:
    """Encode a single VLP v0.2 frame into 32 bytes."""
    if isinstance(status, str):
        status = STATUS_BY_NAME[status]
    status = Status(status)
    head = struct.pack(
        "<2sBBIQQI",
        MAGIC,
        VERSION,
        int(status),
        pid,
        timestamp,
        nonce,
        payload,
    )
    assert len(head) == 28
    return head + struct.pack("<I", crc32c(head))


def decode(buf: bytes) -> Frame:
    """Decode a 32-byte VLP v0.2 frame.

    Raises ``DecodeError`` on the first failed validation step. See
    `book/src/spec/vlp.md` §5 for the normative decode order.
    """
    if len(buf) != 32:
        raise DecodeError("BadMagic", f"length {len(buf)} != 32")
    if buf[0:2] != MAGIC:
        raise DecodeError("BadMagic", buf[0:2].hex())
    if buf[2] != VERSION:
        raise DecodeError("BadVersion", f"0x{buf[2]:02x}")

    stored_crc, = struct.unpack("<I", buf[28:32])
    computed_crc = crc32c(buf[0:28])
    if stored_crc != computed_crc:
        raise DecodeError("BadCrc", f"expected {computed_crc:08x}, got {stored_crc:08x}")

    status_byte = buf[3]
    if status_byte not in (Status.OK, Status.DEGRADED, Status.CRITICAL, Status.STALL):
        raise DecodeError("BadStatus", f"0x{status_byte:02x}")
    if status_byte == Status.STALL:
        raise DecodeError("StallOnWire")

    pid, = struct.unpack("<I", buf[4:8])
    timestamp, = struct.unpack("<Q", buf[8:16])
    nonce, = struct.unpack("<Q", buf[16:24])
    payload, = struct.unpack("<I", buf[24:28])

    if pid in (0, 1):
        raise DecodeError("BadPid", str(pid))
    if timestamp == 0xFFFFFFFFFFFFFFFF:
        raise DecodeError("BadTimestamp")
    if nonce == NONCE_TERMINAL and status_byte != Status.CRITICAL:
        raise DecodeError("BadNonce", f"nonce=NONCE_TERMINAL paired with status=0x{status_byte:02x}")

    return Frame(
        status=Status(status_byte),
        pid=pid,
        timestamp=timestamp,
        nonce=nonce,
        payload=payload,
    )

File: tools/reference-implementations/go/vlp.go
// Package vlp is a non-normative Go reference for VLP v0.2 base frames.
//
// The authoritative specification lives at book/src/spec/vlp.md. This
// package exists so a Go developer can confirm their understanding of
// the wire format against working code without needing a Rust toolchain.
//
// Standard library only. Requires Go 1.21+.
package main

import (
	"encoding/binary"
	"fmt"
)

// Magic prefix, version byte, and reserved nonce sentinel.
var (
	Magic          = [2]byte{0x56, 0x41} // "VA"
	Version        = byte(0x02)
	NonceTerminal  = uint64(0xFFFFFFFFFFFFFFFF)
)

// Status enumerates the agent's reported health.
type Status uint8

const (
	StatusOk       Status = 0
	StatusDegraded Status = 1
	StatusCritical Status = 2
	StatusStall    Status = 3 // observer-synthesized only — MUST NOT appear on the wire
)

func StatusFromName(name string) (Status, error) {
	switch name {
	case "ok":
		return StatusOk, nil
	case "degraded":
		return StatusDegraded, nil
	case "critical":
		return StatusCritical, nil
	default:
		return 0, fmt.Errorf("unknown status name %q", name)
	}
}

// Frame is the decoded form of a 32-byte wire frame.
type Frame struct {
	Status    Status
	PID       uint32
	Timestamp uint64
	Nonce     uint64
	Payload   uint32
}

// DecodeError carries the spec-defined error variant name. The Kind field
// matches the strings in tools/vlp-test-vectors.json.
type DecodeError struct {
	Kind   string
	Detail string
}

func (e *DecodeError) Error() string {
	if e.Detail == "" {
		return e.Kind
	}
	return e.Kind + ": " + e.Detail
}

// ----------------------------------------------------------------------------
// CRC-32C (Castagnoli) — byte-at-a-time table lookup. RFC 3720 appendix B.
// ----------------------------------------------------------------------------

const crc32cReflected = uint32(0x82F63B78)

var crc32cTable = buildCRC32CTable()

func buildCRC32CTable() [256]uint32 {
	var t [256]uint32
	for i := uint32(0); i < 256; i++ {
		c := i
		for j := 0; j < 8; j++ {
			if c&1 != 0 {
				c = (c >> 1) ^ crc32cReflected
			} else {
				c = c >> 1
			}
		}
		t[i] = c
	}
	return t
}

// CRC32C computes the Castagnoli CRC-32C of data.
func CRC32C(data []byte) uint32 {
	crc := uint32(0xFFFFFFFF)
	for _, b := range data {
		crc = crc32cTable[(crc^uint32(b))&0xff] ^ (crc >> 8)
	}
	return crc ^ 0xFFFFFFFF
}

// ----------------------------------------------------------------------------
// Encode / Decode
// ----------------------------------------------------------------------------

// Encode produces a 32-byte VLP v0.2 frame.
func Encode(status Status, pid uint32, timestamp, nonce uint64, payload uint32) [32]byte {
	var out [32]byte
	out[0] = Magic[0]
	out[1] = Magic[1]
	out[2] = Version
	out[3] = byte(status)
	binary.LittleEndian.PutUint32(out[4:8], pid)
	binary.LittleEndian.PutUint64(out[8:16], timestamp)
	binary.LittleEndian.PutUint64(out[16:24], nonce)
	binary.LittleEndian.PutUint32(out[24:28], payload)
	binary.LittleEndian.PutUint32(out[28:32], CRC32C(out[0:28]))
	return out
}

// Decode validates a 32-byte buffer and returns the recovered Frame.
// On any validation failure it returns *DecodeError; see
// book/src/spec/vlp.md §5 for the normative decode order.
func Decode(buf []byte) (Frame, error) {
	if len(buf) != 32 {
		return Frame{}, &DecodeError{Kind: "BadMagic", Detail: fmt.Sprintf("length %d != 32", len(buf))}
	}
	if buf[0] != Magic[0] || buf[1] != Magic[1] {
		return Frame{}, &DecodeError{Kind: "BadMagic", Detail: fmt.Sprintf("%02x%02x", buf[0], buf[1])}
	}
	if buf[2] != Version {
		return Frame{}, &DecodeError{Kind: "BadVersion", Detail: fmt.Sprintf("0x%02x", buf[2])}
	}

	storedCRC := binary.LittleEndian.Uint32(buf[28:32])
	computedCRC := CRC32C(buf[0:28])
	if storedCRC != computedCRC {
		return Frame{}, &DecodeError{
			Kind:   "BadCrc",
			Detail: fmt.Sprintf("expected %08x, got %08x", computedCRC, storedCRC),
		}
	}

	statusByte := buf[3]
	if statusByte > 3 {
		return Frame{}, &DecodeError{Kind: "BadStatus", Detail: fmt.Sprintf("0x%02x", statusByte)}
	}
	if statusByte == byte(StatusStall) {
		return Frame{}, &DecodeError{Kind: "StallOnWire"}
	}

	pid := binary.LittleEndian.Uint32(buf[4:8])
	ts := binary.LittleEndian.Uint64(buf[8:16])
	nonce := binary.LittleEndian.Uint64(buf[16:24])
	payload := binary.LittleEndian.Uint32(buf[24:28])

	if pid == 0 || pid == 1 {
		return Frame{}, &DecodeError{Kind: "BadPid", Detail: fmt.Sprintf("%d", pid)}
	}
	if ts == 0xFFFFFFFFFFFFFFFF {
		return Frame{}, &DecodeError{Kind: "BadTimestamp"}
	}
	if nonce == NonceTerminal && statusByte != byte(StatusCritical) {
		return Frame{}, &DecodeError{
			Kind:   "BadNonce",
			Detail: fmt.Sprintf("status 0x%02x with terminal nonce", statusByte),
		}
	}

	return Frame{
		Status:    Status(statusByte),
		PID:       pid,
		Timestamp: ts,
		Nonce:     nonce,
		Payload:   payload,
	}, nil
}
